server/test/test_repository.go

```diff
@@ -0,0 +1,236 @@
+
+	const condition = " AND (t.name ilike $3 OR t.description ilike $3)"
+	if hasSearchQuery {
+		params = append(params, cleanSearchQuery)
+		sql += condition
+	}
+
+	sortingFields := map[string]string{
+		"created":  "t.created_at",
+		"name":     "t.name",
+		"last_run": "last_test_run_time",
+	}
+
+	sql = sortQuery(sql, sortBy, sortDirection, sortingFields)
+	sql += ` LIMIT $1 OFFSET $2 `
+
+	stmt, err := r.db.Prepare(sql)
+	if err != nil {
+		return []Test{}, err
+	}
+	defer stmt.Close()
+
+	rows, err := stmt.QueryContext(ctx, params...)
+	if err != nil {
+		return []Test{}, err
+	}
+
+	tests, err := r.readTestRows(ctx, rows)
+	if err != nil {
+		return []Test{}, err
+	}
+
+	return tests, nil
+}
+
+func sortQuery(sql, sortBy, sortDirection string, sortingFields map[string]string) string {
+	sortField, ok := sortingFields[sortBy]
+
+	if !ok {
+		sortField = sortingFields["created"]
+	}
+
+	dir := "DESC"
+	if strings.ToLower(sortDirection) == "asc" {
+		dir = "ASC"
+	}
+
+	return fmt.Sprintf("%s ORDER BY %s %s", sql, sortField, dir)
+}
+
+func (r *repository) Count(ctx context.Context, query string) (int, error) {
+	countQuery := "SELECT COUNT(*) FROM tests t" + testMaxVersionQuery
+
+	if query != "" {
+		countQuery = fmt.Sprintf("%s WHERE %s", countQuery, query)
+	}
+
+	count := 0
+
+	err := r.db.
+		QueryRowContext(ctx, countQuery).
+		Scan(&count)
+
+	if err != nil {
+		return 0, fmt.Errorf("sql query: %w", err)
+	}
+
+	return count, nil
+}
+
+type scanner interface {
+	Scan(dest ...interface{}) error
+}
+
+func (r *repository) readTestRows(ctx context.Context, rows *sql.Rows) ([]Test, error) {
+	tests := []Test{}
+
+	for rows.Next() {
+		test, err := r.readTestRow(ctx, rows)
+		if err != nil {
+			return []Test{}, err
+		}
+
+		tests = append(tests, test)
+	}
+
+	return tests, nil
+}
+
+func (r *repository) readTestRow(ctx context.Context, row scanner) (Test, error) {
+	test := Test{}
+
+	var (
+		jsonServiceUnderTest,
+		jsonSpecs,
+		jsonOutputs []byte
+
+		lastRunTime *time.Time
+
+		pass, fail *int
+	)
+	err := row.Scan(
+		&test.ID,
+		&test.Version,
+		&test.Name,
+		&test.Description,
+		&jsonServiceUnderTest,
+		&jsonSpecs,
+		&jsonOutputs,
+		&test.CreatedAt,
+		&test.Summary.Runs,
+		&lastRunTime,
+		&pass,
+		&fail,
+	)
+
+	switch err {
+	case nil:
+		err = json.Unmarshal(jsonServiceUnderTest, &test.Trigger)
+		if err != nil {
+			return Test{}, fmt.Errorf("cannot parse trigger: %w", err)
+		}
+
+		err = json.Unmarshal(jsonSpecs, &test.Specs)
+		if err != nil {
+			return Test{}, fmt.Errorf("cannot parse specs: %w", err)
+		}
+
+		err = json.Unmarshal(jsonOutputs, &test.Outputs)
+		if err != nil {
+			return Test{}, fmt.Errorf("cannot parse outputs: %w", err)
+		}
+
+		if lastRunTime != nil {
+			test.Summary.LastRun.Time = *lastRunTime
+		}
+		if pass != nil {
+			test.Summary.LastRun.Passes = *pass
+		}
+		if fail != nil {
+			test.Summary.LastRun.Fails = *fail
+		}
+
+		return test, nil
+	default:
+		return Test{}, err
+	}
```

Comment:
since we don't check for the error type anymore, we can remove the switch and make the code more "standard" looking:

```suggestion
	if err != nil {
		return Test{}, fmt.Errorf("cannot read row: %w", err)
	}

	err = json.Unmarshal(jsonServiceUnderTest, &test.Trigger)
	if err != nil {
		return Test{}, fmt.Errorf("cannot parse trigger: %w", err)
	}

	err = json.Unmarshal(jsonSpecs, &test.Specs)
	if err != nil {
		return Test{}, fmt.Errorf("cannot parse specs: %w", err)
	}

	err = json.Unmarshal(jsonOutputs, &test.Outputs)
	if err != nil {
		return Test{}, fmt.Errorf("cannot parse outputs: %w", err)
	}

	if lastRunTime != nil {
		test.Summary.LastRun.Time = *lastRunTime
	}
	if pass != nil {
		test.Summary.LastRun.Passes = *pass
	}
	if fail != nil {
		test.Summary.LastRun.Fails = *fail
	}

	return test, nil
```